tests/rust/safe_abstraction/mutex_u32.rs

```diff
@@ -0,0 +1,444 @@
+
+pred_ctor MutexU32_frac_borrow_content(kfcc: lifetime_t, t: thread_id_t, l: *MutexU32)(;) =
+     SysMutex_share(&(*l).inner, full_borrow_wrapper(kfcc, u32_full_borrow_content(t, &(*l).data)));
+
+pred MutexU32_share(k: lifetime_t, t: thread_id_t, l: *MutexU32) =
+    exists_np(?kfcc) &*& lifetime_inclusion(k, kfcc) == true &*& frac_borrow(k, MutexU32_frac_borrow_content(kfcc, t, l));
+
+lem MutexU32_share_mono(k: lifetime_t, k1: lifetime_t, t: thread_id_t, l: *MutexU32)
+    req lifetime_inclusion(k1, k) == true &*& [_]MutexU32_share(k, t, l);
+    ens [_]MutexU32_share(k1, t, l);
+{
+    open MutexU32_share(k, t, l); assert [_]exists_np(?kfcc);
+    frac_borrow_mono(k, k1, MutexU32_frac_borrow_content(kfcc, t, l));
+    assert [?q]frac_borrow(k1, _); close [q]exists_np(kfcc);
+    // Todo: Why does VeriFast not just close using any dummy fraction when it is trying to close a dummy fraction?
+    lifetime_inclusion_trans(k1, k, kfcc);
+    close [q]MutexU32_share(k1, t, l);
+}
+
+lem MutexU32_share_full(k: lifetime_t, t: thread_id_t, l: *MutexU32)
+    req atomic_mask(Nlft) &*& full_borrow(k, MutexU32_full_borrow_content0(t, l)) &*& [?q]lifetime_token(k);
+    ens atomic_mask(Nlft) &*& [_]MutexU32_share(k, t, l) &*& [q]lifetime_token(k);
+{
+
+    produce_lem_ptr_chunk implies(sep(MutexU32_fbc_inner(l), u32_full_borrow_content(t, &(*l).data)), MutexU32_full_borrow_content0(t, l))() {
+        open sep(MutexU32_fbc_inner(l), u32_full_borrow_content(t, &(*l).data))();
+        open MutexU32_fbc_inner(l)(); open u32_full_borrow_content(t, &(*l).data)();
+        assert (*l).inner |-> ?inner &*& (*l).data |-> ?data;
+        close MutexU32_own(t, inner, data);
+        close MutexU32_full_borrow_content0(t, l)();
+    }{
+        produce_lem_ptr_chunk implies(MutexU32_full_borrow_content0(t, l), sep(MutexU32_fbc_inner(l), u32_full_borrow_content(t, &(*l).data)))() {
+            open MutexU32_full_borrow_content0(t, l)();
+            assert (*l).inner |-> ?inner &*& (*l).data |-> ?data;
+            open MutexU32_own(t, inner, data);
+            close MutexU32_fbc_inner(l)();
+            close u32_full_borrow_content(t, &(*l).data)();
+            close sep(MutexU32_fbc_inner(l), u32_full_borrow_content(t, &(*l).data))();
+        }{
+            full_borrow_implies(k, MutexU32_full_borrow_content0(t, l), sep(MutexU32_fbc_inner(l), u32_full_borrow_content(t, &(*l).data)));
+        }
+        full_borrow_split_m(k, MutexU32_fbc_inner(l), u32_full_borrow_content(t, &(*l).data));
+        let kstrong = open_full_borrow_strong_m(k, MutexU32_fbc_inner(l), q);
+        produce_lem_ptr_chunk full_borrow_convert_strong(True, MutexU32_frac_borrow_content(k, t, l), kstrong, MutexU32_fbc_inner(l))() {
+            open MutexU32_frac_borrow_content(k, t, l)();
+            SysMutex_end_share(&(*l).inner);
+            assert (*l).inner |-> ?inner;
+            SysMutex_renew(inner, True);
```

Comment:
Very good point. I do not think these lemmas are provable at the bottom mask. I think the least Iris feature necessary for this will be invariant, actually a cancellative one.
`full_borrow_conovert_strong` should be proven at the bottom mask. Does that mean it is broken?
Another concern that comes to my mind is the `Ctx()` predicate I have added to `full_borrow_convert_strong`'s contract. One could use it to sneak a mask token in. It does not give the `Ctx()` back in `close_full_borrow_strong`'s post-condition. So it should be fine but thought it worth mentioning to you.

Reply:
To be investigated at some point in the future (probably after we implement SysMutex)...